gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -116,139 +132,103 @@ public AbstractJobLauncher(Properties properties)
       taskExecutor.submit(task);
     }
 
-    LOG.info(String.format("Waiting for submitted tasks of job %s to complete...", jobId));
-    while (countDownLatch.getCount() > 0) {
-      LOG.info(String
-          .format("%d out of %d tasks of job %s are running", countDownLatch.getCount(), workUnits.size(), jobId));
-      countDownLatch.await(1, TimeUnit.MINUTES);
-    }
-    LOG.info(String.format("All tasks of job %s have completed", jobId));
-
     return tasks;
   }
 
+  /**
+   * This implementation simply sets a flag that indicate cancellation is requested and waits
+   * for the cancellation to be executed. It does not guarantee successful job cancellation.
+   *
+   * {@inheritDoc JobLauncher#cancelJob(JobListener)}
+   */
   @Override
-  @SuppressWarnings("unchecked")
-  public void launchJob(Properties jobProps, JobListener jobListener)
+  public void cancelJob(JobListener jobListener)
```

Comment:
* What is the current contract of the cancelJob() API? Is it synchronous or asynchronous? Are we guaranteeing to always call the listener? 

Currently, it seems like sometimes the call is executed synchronously and sometimes, it is not. The listener is called back only if the call is executed synchronously but I think that kind of defeats the purpose of having a listener (aka callback). If the call is synchronous, I can execute whatever I want after the cancelJob() call.
I think the current form of the API suggests asynchronous cancellation. This means that the implementation should always return immediately and always call the callback regardless if there is cancellation in progress or not. We can of course have different contract but it should be described very carefully in the javadocs.

* How does cancelJob() relate to the state property in the job state? My intuition is that cancelJob() should set the state to CANCELLED at the end of the run but that does not seem to be the case currently. The state property is managed independently which I think is error prone. I may be missing some interaction, in which case it may be useful to document it properly. 

* Overall, I think we need a global lock on state transitions, not only cancellationLock and we need to manage those more carefully. Currently, close() seems to be closing the cancellation executor regardless of whether a cancellation is in progress. Similarly, I think it is also possible to cancel a job after it has already finished (successfully or not). I find this counter-intuitive. I think having a state machine diagram with the transition will be helpful.

* I think the current implementation with synchronization on cancellationExecutor and cancellationLock is a bit complicated. I think having a single state transition lock (see above) and appropriate condition variables might be cleaner.